File: javascript/fibonacci.js
//O(n^2)
var fib = function(num) {
  if (num <= 1) {
    return 1;
  }

  return fib(num - 1) + fib(num - 2)
}

//O(n)
var fibonacci = function(num) {
  var a = 1, b = 0, temp;
  while (num >= 0) {
    temp = a;
    a = a + b;
    b = temp;
    num--;
  }
  return b
}

//0(n) for space, 2N for time
var fibo = function(num, mem) {
  mem = mem || {};

  if (mem[num]) return mem[num];
  if (num <= 1) return 1;

  return mem[num] = fibo(num - 1, mem) + fibo(num - 2, mem)
}

console.log(fib(6))
console.log(fibonacci(6))
console.log(fibo(6))

File: ruby/fibonacci.rb
def fib(n)
  if n <= 2
    return 1
  else
    return fib(n-1) + fib(n-2)
  end
end

def fibo(n)
  index = 3
  a, b = 1, 1
  if n <= 2
    return 1
  else
    while index <= n do
      temp = a + b
      a = b
      b = temp
      index += 1
    end
  end
  return temp
end
